File: docs/manual/java/guide/cluster/code/docs/home/persistence/Post2.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.persistence;

import docs.home.persistence.BlogCommand.*;
import docs.home.persistence.BlogEvent.*;
import java.util.Optional;
import akka.Done;

// #post1
import com.lightbend.lagom.javadsl.persistence.PersistentEntity;

public class Post2 extends PersistentEntity<BlogCommand, BlogEvent, BlogState> {

  @Override
  public Behavior initialBehavior(Optional<BlogState> snapshotState) {
    BehaviorBuilder b = newBehaviorBuilder(snapshotState.orElse(BlogState.EMPTY));

    // #command-handler
    // Command handlers are invoked for incoming messages (commands).
    // A command handler must "return" the events to be persisted (if any).
    b.setCommandHandler(
        AddPost.class,
        (AddPost cmd, CommandContext<AddPostDone> ctx) -> {
          final PostAdded postAdded = new PostAdded(entityId(), cmd.getContent());
          return ctx.thenPersist(
              postAdded,
              (PostAdded evt) ->
                  // After persist is done additional side effects can be performed
                  ctx.reply(new AddPostDone(entityId())));
        });
    // #command-handler

    // #validate-command
    b.setCommandHandler(
        AddPost.class,
        (AddPost cmd, CommandContext<AddPostDone> ctx) -> {
          if (cmd.getContent().getTitle() == null || cmd.getContent().getTitle().equals("")) {
            ctx.invalidCommand("Title must be defined");
            return ctx.done();
          }
          // #validate-command

          final PostAdded postAdded = new PostAdded(entityId(), cmd.getContent());
          return ctx.thenPersist(
              postAdded,
              (PostAdded evt) ->
                  // After persist is done additional side effects can be performed
                  ctx.reply(new AddPostDone(entityId())));
        });

    // #event-handler
    // Event handlers are used both when persisting new events
    // and when replaying events.
    b.setEventHandler(PostAdded.class, evt -> new BlogState(Optional.of(evt.getContent()), false));
    // #event-handler

    // #change-behavior
    b.setEventHandlerChangingBehavior(
        PostAdded.class,
        evt -> becomePostAdded(new BlogState(Optional.of(evt.getContent()), false)));
    // #change-behavior

    return b.build();
  }

  // #change-behavior-become
  private Behavior becomePostAdded(BlogState newState) {
    BehaviorBuilder b = newBehaviorBuilder(newState);

    b.setReadOnlyCommandHandler(GetPost.class, (cmd, ctx) -> ctx.reply(state().getContent().get()));

    b.setCommandHandler(
        ChangeBody.class,
        (cmd, ctx) ->
            ctx.thenPersist(
                new BodyChanged(entityId(), cmd.getBody()), evt -> ctx.reply(Done.getInstance())));

    b.setEventHandler(BodyChanged.class, evt -> state().withBody(evt.getBody()));

    return b.build();
  }
  // #change-behavior-become

}

File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/Post2.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

import akka.Done
import com.lightbend.lagom.scaladsl.persistence.AggregateEvent
import com.lightbend.lagom.scaladsl.persistence.AggregateEventShards
import com.lightbend.lagom.scaladsl.persistence.AggregateEventTag
import com.lightbend.lagom.scaladsl.persistence.PersistentEntity

import com.lightbend.lagom.scaladsl.persistence.PersistentEntity.ReplyType

final class Post2 extends PersistentEntity {

  override type Command = BlogCommand
  override type Event   = BlogEvent
  override type State   = BlogState

  override def initialState: BlogState = BlogState.empty

  //#behavior
  override def behavior: Behavior = {
    case state if state.isEmpty  => initial
    case state if !state.isEmpty => postAdded
  }
  //#behavior

  //#initial-actions
  private val initial: Actions = {
    //#initial-actions
    //#command-handler
    // Command handlers are invoked for incoming messages (commands).
    // A command handler must "return" the events to be persisted (if any).
    Actions()
      .onCommand[AddPost, AddPostDone] {
        case (AddPost(content), ctx, state) =>
          ctx.thenPersist(PostAdded(entityId, content)) { evt =>
            // After persist is done additional side effects can be performed
            ctx.reply(AddPostDone(entityId))
          }
      }
      //#command-handler
      //#validate-command
      // Command handlers are invoked for incoming messages (commands).
      // A command handler must "return" the events to be persisted (if any).
      .onCommand[AddPost, AddPostDone] {
        case (AddPost(content), ctx, state) =>
          if (content.title == null || content.title.equals("")) {
            ctx.invalidCommand("Title must be defined")
            ctx.done
          }
          //#validate-command
          else {
            ctx.thenPersist(PostAdded(entityId, content)) { _ =>
              // After persist is done additional side effects can be performed
              ctx.reply(AddPostDone(entityId))
            }
          }
      }
      //#validate-command
      //#event-handler
      // Event handlers are used both when persisting new events and when replaying
      // events.
      .onEvent {
        case (PostAdded(postId, content), state) =>
          BlogState(Some(content), published = false)
      }
    //#event-handler
  }

  //#postAdded-actions
  private val postAdded: Actions = {
    //#postAdded-actions
    Actions()
    //#reply
      .onCommand[ChangeBody, Done] {
        case (ChangeBody(body), ctx, state) =>
          ctx.thenPersist(BodyChanged(entityId, body))(_ => ctx.reply(Done))
      }
      //#reply
      .onEvent {
        case (BodyChanged(_, body), state) =>
          state.withBody(body)
      }
      //#read-only-command-handler
      .onReadOnlyCommand[GetPost.type, PostContent] {
        case (GetPost, ctx, state) =>
          ctx.reply(state.content.get)
      }
    //#read-only-command-handler
  }

}
